src/engine/renderer/particle_system.cpp

```diff
@@ -60,20 +66,31 @@ Particle ParticleSystem::generate() {
             damping.random(),
 
             first_size,
+            first_size_deriv,
             second_size,
-
+            second_size_deriv,
+            &progress_func_size,
             V2(width.random(), height.random()),
 
             first_color,
+            first_color_deriv,
             second_color,
+            second_color_deriv,
+            &progress_func_color,
             (s16) (num_sub_sprites ?  random_int() % num_sub_sprites : -1),
     };
 }
 
 void ParticleSystem::spawn(u32 num_particles) {
     ASSERT(particles, "Trying to use uninitalized/destroyed particle system");
     for (u32 i = 0; i < num_particles; i++) {
-        if (head == tail) return;
+        if (head == tail) {
+            if (!drop_oldest) {
+                return;
+            } else {
+                head = (head + 1) % max_num_particles;
+            }
```

Comment:
I personally drop the else statement here, since it can be thought of as the code continuing. But that's just a matter of taste.